File: web/tests/reference/sliding-window/minimum-window-substring.py
def min_window(s, t):
    if not t or not s:
        return ""

    need = {}
    for ch in t:
        need[ch] = need.get(ch, 0) + 1
    required = len(need)

    have = {}
    satisfied = 0

    best_len = float("inf")
    best_left = 0
    left = 0

    for right, ch in enumerate(s):
        have[ch] = have.get(ch, 0) + 1
        if ch in need and have[ch] == need[ch]:
            satisfied += 1

        while satisfied == required:
            if right - left + 1 < best_len:
                best_len = right - left + 1
                best_left = left

            left_ch = s[left]
            have[left_ch] -= 1
            if left_ch in need and have[left_ch] < need[left_ch]:
                satisfied -= 1
            left += 1

    return "" if best_len == float("inf") else s[best_left : best_left + best_len]

File: web/tests/reference/sliding-window/minimum-window-substring.js
function minWindow(s, t) {
  if (!t || !s) return "";

  const need = new Map();
  for (const ch of t) need.set(ch, (need.get(ch) ?? 0) + 1);
  const required = need.size;

  const have = new Map();
  let satisfied = 0;

  let bestLen = Infinity;
  let bestLeft = 0;
  let left = 0;

  for (let right = 0; right < s.length; right++) {
    const ch = s[right];
    have.set(ch, (have.get(ch) ?? 0) + 1);
    if (need.has(ch) && have.get(ch) === need.get(ch)) satisfied++;

    while (satisfied === required) {
      if (right - left + 1 < bestLen) {
        bestLen = right - left + 1;
        bestLeft = left;
      }

      const leftCh = s[left];
      have.set(leftCh, have.get(leftCh) - 1);
      if (need.has(leftCh) && have.get(leftCh) < need.get(leftCh)) satisfied--;
      left++;
    }
  }

  return bestLen === Infinity ? "" : s.slice(bestLeft, bestLeft + bestLen);
}
